### PR TITLE
Make the soure generator format enums using their identifiers rather than numeric values.

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Updates the enum literal value formatter so that they are specified using their names rather than numeric values. For example the source generated code:
```C#
NumberHandling = (global::System.Text.Json.Serialization.JsonNumberHandling)3,
```
is now changed to
```C#
NumberHandling = global::System.Text.Json.Serialization.JsonNumberHandling.AllowReadingFromString | global::System.Text.Json.Serialization.JsonNumberHandling.WriteAsString,
```

Contributes to #79534.

### CODE REVIEW

File: src/libraries/System.Text.Json/gen/Helpers/EnumFormatter.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json.SourceGeneration
+{
+    public static class EnumFormatter<TEnum> where TEnum : struct, Enum
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/gen/Helpers/EnumFormatter.cs:12]: This is a ton of code duplicating logic in Enum itself.  Can you instead just do something like:
```C#
static string FormatEnum<TEnum>(TEnum value) where TEnum : Enum
{
    string[] values = value.ToString().Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
    return string.Join(" | ", values.Select(s =>
        char.IsAsciiDigit(s[0]) ?
            $"({typeof(TEnum).Name}){s}" :
            $"{typeof(TEnum).Name}.{s}"));
}
```
?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/Helpers/EnumFormatter.cs:12]: I could, but I figured doing that was preferable to having to parse the `.ToString()` output for fearing of missing a corner case.

Comment (stephentoub) [reply src/libraries/System.Text.Json/gen/Helpers/EnumFormatter.cs:12]: The code here also misses corner cases.  For example, this:
```C#
Console.WriteLine(FormatEnum(AttributeTargets.All));
Console.WriteLine();
Console.WriteLine(EnumFormatter<AttributeTargets>.FormatEnumLiteral(nameof(AttributeTargets), AttributeTargets.All));
```
produces this:
```
AttributeTargets.All

AttributeTargets.Assembly | AttributeTargets.Module | AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Enum | AttributeTargets.Constructor | AttributeTargets.Method | AttributeTargets.Property | AttributeTargets.Field | AttributeTargets.Event | AttributeTargets.Interface | AttributeTargets.Parameter | AttributeTargets.Delegate | AttributeTargets.ReturnValue | AttributeTargets.GenericParameter
```
I'd rather have the formatting be done by the tried-and-true implementation than try to handle all the various cases the Enum implementation already handles.  We know what Enum.ToString can produce, and we can have tests for various cases.  If we encounter something really remote that we somehow missed, we can fix it.

My real hesitancy is that this is both a lot of logic while still being incomplete.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/Helpers/EnumFormatter.cs]: I've updated the PR using a variation of your proposed solution.

**Review by stephentoub — APPROVED** [review]

**Pull request merged**